<p style="text-align: left;"> Uğur Erdem Seyfi <br> 21801744 </p>


# <p style="text-align: center;"> MATH420 HW02 </p>


# 2.5

**Question:** Let $p$ be an odd prime and let $g$ be a primitive root modulo $p$. Prove that $a$ has a square root modulo $p$ if and only if its discrete logarithm $log_g(a)$ modulo $p−1$ is even.

**Solution:**

Now $p$ be an odd prime and let $g$ be a primitive root modulo $p$.
We want to prove that $a$ has a square root modulo $p$ if and only if its discrete logarithm $log_g(a)$ modulo $p−1$ is even.

**Proof of only if part:**

Assume that $a$ has a square root module $p$. This means that there exist an integer $b$ such that $b^2 \equiv a \text{ (mod p) }$  

Then, let $k = log_g(b) \text{ (mod p)}$, for $ k \in [0, p-1]$. Then we have $ g^{2k} = b^2 = a$ Let $ 2k=n(p−1)+m$ for $ n \in \mathbb{Z}$ and $ m \in [0, p-1]$ Then we have $ g^m = g^{2k−n(p−1)} = \frac{g^{2k} }{g^{n(p-1)}} $ which is equal to $ g^{2k} $ since $g^{n(p-1)}$ is equal to 1 by FTL. This implies $ g^{2k} = a \text{ (mod p)}$, that is to say, $log_g(a) = m = 2k$, hence $log_g(a)$ is even.


**Proof of if part:**

Let p be an odd prime and let g be a primitive root modulo p. Let $log_g(a) \text{ (mod p−1) }$ be even. Then, $g^{2k} = a$, for some $k \in \mathbb{Z} $. This implies that there exist a b such that $b^2 = a$, specifically for $b = g^k$. Hence, a has a square root modulo p.


# 2.6

**Question:** Alice and Bob agree to use the prime $p = 1373$ and the base $g = 2$ for a Diffie–Hellman key exchange. Alice sends Bob the value $A = 974$. Bob asks your assistance, so you tell him to use the secret exponent $b = 871$. What value $B$ should Bob send to Alice, and what is their secret shared value? Can you figure out Alice’s secret exponent?

**Solution:**

Bob sends $B = 805$ since we have $ g^b \equiv 2^{871} \equiv 805 \text{ (mod 1373)}$ 

Their secret shared value is $ A^{b} \equiv 974^{871} \equiv 397 \text{ (mod 1373) }$ 

We can solve Alice's secret exponent by using the Shank's Babystep-Giantstep Algorithm. 



In [2]:
from sage.groups.generic import bsgs

g = Mod(2, 1373)
A = Mod(974, 1373)
print(" Solution to the DLP is, x = ", bsgs(g, A, (0, 1373)))

 Solution to the DLP is, x =  587


so Alice's secret exponent is $ a = 587 $.


# 2.8

**Question:** Alice and Bob agree to use the prime p = 1373 and the base g = 2 for communications using the Elgamal public key cryptosystem.

**Q(a)**: Alice chooses a = 947 as her private key. What is the value of her public key A?

**Solution:** $ A = g^a = 2^{947} = 177 \text{ (mod 1373) }$ 


**Q(b)**: Bob chooses $b = 716$ as his private key, so his public key is $B \equiv 2^{716} \equiv 469 \text{(mod 1373)}$. Alice encrypts the message $m = 583$ using the random element $k = 877$. What is the ciphertext $(c1, c2)$ that Alice sends to Bob?

**Solution:** 

We have $ c_1 = 2^{877} = 719 \text{ (mod 1373) }$ and $ c_2 = 583 * 469^{877} = 623 \text{ (mod 1373) } $ so Alice sends $(c_1, c_2) = (719, 623)$ to Bob.

**Q(c)**: Alice decides to choose a new private key $a = 299$ with associated public key $ A \equiv 2299 \equiv 34 \text{ (mod 1373) }$. Bob encrypts a message using Alice’s public key and sends her the ciphertext $(c1, c2) = (661, 1325)$. Decrypt the message.

**Solution:** 
let $K = 661^{299} = 645 \text{ (mod 1373)}$ denote our shared key, then we can use $ d = K^{-1} \text{ (mod 1373)}$ to multiply with $c_2$ and recover our original message $m$. By Fermat's Little Theorem we have $ d = K^{-1} = 645^{1371} = 794 \text{ (mod 1373)} $ and thus we have $ m = dc_2 = 794.1325 = 332 \text{ (mod 1373)} $ is our message.

**Q(d)**: Now Bob chooses a new private key and publishes the associated public key $B = 893$. Alice encrypts a message using this public key and sends the ciphertext $(c1, c2) = (693, 793)$ to Bob. Eve intercepts the transmission. Help Eve by solving the discrete logarithm problem $2^b \equiv 893 \text{ (mod 1373)}$ and using the value of b to decrypt the message.

**Solution:**

We can find the exponent b by using Shank's Babystep-Giantstep Algorithm.

In [3]:
g = Mod(2, 1373)
A = Mod(893, 1373)
print(" Solution to the DLP is, b = ", bsgs(g, A, (0, 1373)))

 Solution to the DLP is, b =  219


In [4]:
693^219 % 1373

431

now since by solving the DLP we obtained $ b = 219 $, 

now since $ B^a = A^b$ we should have $B^a \equiv 693^219 \equiv 431 \text{ (mod 1373) }$. And since $ c_2 \equiv A^b m \text{ mod(1373)}$ it is enough for us to find inverse of 431 modulo 1373.



In [5]:
print("inverse 431 modulo 1373 is: ", inverse_mod(431, 1373))

inverse 431 modulo 1373 is:  532


Thus our message is $m = 532$


# 2.10

**Question:**

The exercise describes a public key cryptosystem that requires Bob and Alice to exchange several messages. We illustrate the system with an example. Bob and Alice fix a publicly known prime p = 32611, and all of the other numbers used are private. Alice takes her message m = 11111, chooses a random exponent a = 3589, and sends the number u = ma (mod p) = 15950 to Bob. Bob chooses a random exponent b = 4037 and sends v = ub (mod p) = 15422 back to Alice. Alice
then computes w = v15619 ≡ 27257 (mod 32611) and sends w = 27257 to Bob. Finally, Bob computes w31883 (mod 32611) and recovers the value 11111 of Alice’s message.

**(a)** Explain why this algorithm works. In particular, Alice uses the numbers
a = 3589 and 15619 as exponents. How are they related? Similarly, how are
Bob’s exponents b = 4037 and 31883 related?

**Solution:**

The algorithm works because of the formulation that is given in part b.

The numbers are related in the following way:

$$ 15619a \equiv 1 \text{ (mod 32610)} $$
$$ 31883b \equiv 1 \text{ (mod 32610)} $$


**(b)** Formulate a general version of this cryptosystem, i.e., using variables, and show
that it works in general.

**Solution:**

Let's fix a prime number $p$. 

Now, alice chooses her text $m$ and a random exponent $a$ such that gcd(a, p-1) = 1. After this, she sends $ u = m^a \text{ (mod p)}$ to Bob. 

Then, Bob chooses a random exponent $b$ such that gcd(b, p-1) = 1. He then calculates $ v = u^b \text{ (mod p) }$, he send v back to Alice.

Now Alice, knowing a, computes the inverse of a modulo p-1. And then does the following calculation: $ w = v^{a^{-1}} \text{ (mod p)}$ and sends w back to Bob.

And finally Bob, computers b inverse modulo p-1. And then does the following calculation to find the original message m: 
$$ w^{b^{-1}} = v{a^{-1} b^{-1}} = u^{b a^{-1} b^{-1}} = m^{ab a^{-1} b^{-1}} = m \text{ (mod p) }$$

Hence, we are done.

**(c)** What is the disadvantage of this cryptosystem over Elgamal? (Hint. How many
times must Alice and Bob exchange data?)

**Solution:**

In Elgamal, it sufficient for Alice to send Bob one message, however in this cryptosystem, Alice has to send two messages to Bob, and Bob also needs to send one message back to Alice. So this system requires more time in terms of sending and recieving messages.


**(d)** Are there any advantages of this cryptosystem over Elgamal? In particular, can Eve break it if she can solve the discrete logarithm problem? Can Eve break it if she can solve the Diffie–Hellman problem?

**Solution:**

In this crypto-system, Alice and Bob reveals less information than in ElGamal. Now Eve can break this cryptosystem if she knows how to solve DLP.

Why? Suppose Eve knows how to solve DLP, then she can solve $w \equiv v^x \text{ (mod p)}$, and find $ x = a^{-1} $, then she can find m by calculating $ u^{a^{-1}} = m^{aa^{-1}} = m \text{ (mod p)}$.

Overall we can say that this system is more secure than Elgamal in case it turns out that Diffie-Hellman problem is easier than DLP. 

# 2.17(c)

**Question:** Use Shanks’s babystep–giantstep method to solve the following discrete logarithm problem.

$$ 650^x = 2213 \text{ in } \mathbb{F}_{3571}$$

**Solution:**

We are going to use the algorithm for Shank's babystep–giantstep method to solve this problem. 

In [6]:
a = Mod(650, 3571)
b = Mod(2213, 3571)
print(" Solution to the DLP is, x = ", bsgs(a, b, (0, 3570)))

 Solution to the DLP is, x =  319


So we have $ x = 319 $ as our solution. 

# 2.25

**Question:**
Suppose $n = pq$ with $p$ and $q$ distinct odd primes.

**(a)** Suppose that $gcd(a, pq) = 1$. Prove that if the equation $x^2 \equiv a \text{ (mod n)}$ has any solutions, then it has four solutions.

**Solution:**

Say $n = pq$ with $p, q$ distinct odd primes. Now say that $gcd(a, pq) = 1$  and  $x^2 \equiv a \text{ (mod n)}$ has at least on solution. Then, we want to show that $x^2  \equiv a \text{ (mod n)} $ has at least four solutions.

Now, we know by chinese remainder theorem we can solve the following system of equations:

$$ x \equiv \pm a \text{ (mod p)} $$
$$ x \equiv \pm a \text{ (mod q)} $$

Since we know that such x exists, depending on how we choose the sign of $a$ in each equation, we will arrive at 4 different solutions $x_1, x_2, x_3, x_4$. 

Since these $x_i$'s are inequivalent modulo p or q, we have that they are inequivalent modulo pq, hence all four give rise to distinct square roots modulo pq, as desired

**(b)** Suppose that you had a machine that could find all four solutions for some given $a$. How could you use this machine to factor $n$?

**Solution:**

Lets suppose that this machine gave us all 4 solutions $x_1, x_2, x_3, x_4$, then, taking any two distinct numbers $x_i, x_j$ such that their sum is $s = n_1 + n_2 $. We have that s is not 0 modulo n, we have this sum s is divisible by either p or by q such that $gcd(n,s) = p$ or $gcd(n,s) = q$, which are the factors of n. Therefore, we can find the factors of n by giving the machine gcd(n,s).

So we can just compute gcd(n, s) to get p or q.

# 3.7

**Question:** Alice publishes her RSA public key: modulus $N = 2038667$ and exponent $e = 103$.

**(a)** Bob wants to send Alice the message $m = 892383$. What ciphertext does Bob
send to Alice?

**Solution:** $ m^e \equiv 892383^{103} \equiv 45293 \text { (mod 2038667)} $. So he would send $ c = 45293 $.

**(b)** Alice knows that her modulus factors into a product of two primes, one of which
is p = 1301. Find a decryption exponent d for Alice.

**Solution:** Let's first find our second prime q, since N is composed by two primes we should have $ q =  \frac{N}{p} = \frac{2038667}{1301} = 1567 $ 

Now we want to find $d$ such that $ m^{ed} = m \text{ mod N} $, thus we should have $ ed = k\phi(N) + 1 = k\phi(p)\phi(k) + 1 $ for some k. This is equivalent to

$$ ed = 1 \text{ (mod (1301-1)(1567-1) ) }$$

so we need to find $d$ mod 2035800 where we have 

$$ 103d \equiv 1 \text{ (mod 2035800)} $$


In [1]:
print("inverse of 103 modulo 2035800 is: ", inverse_mod (103, 2035800) )

inverse of 103 modulo 2035800 is:  810367


Thus we have $$ d = 810367 $$

**(c)** Alice receives the ciphertext c = 317730 from Bob. Decrypt the message.

**Solution:** 

$$ c^d \equiv 317730^{810367} \equiv 514407 \text{ (mod N)}$$

so we have $$ m = 514407 $$


# 3.8

**Question:** Bob’s RSA public key has modulus N = 12191 and exponent e = 37. Alice sends Bob the ciphertext c = 587. Unfortunately, Bob has chosen too small a modulus. Help Eve by factoring N and decrypting Alice’s message. (Hint. N has a factor smaller than 100.)

**Solution**: By Brute Force, Eve realizes that N = 73 * 167 = 12191, 

Now we want to find d such that $ed = 1 \text{ mod } \phi(N)$, thus we have $ 37d = 1 \text{ mod 72 * 166}$. By solving this, we find $ d = 11629$

Thus our message is $ m = c^d = 4894 \text{ mod N} $

# 3.10

**Question:**

A decryption exponent for an RSA public key $(N, e)$ is an integer d with the property that $a^{de} \equiv a \text{ (mod N) }$ for all integers a that are relatively prime to $N$.

**(a)** Suppose that Eve has a magic box that creates decryption exponents for (N, e)
for a fixed modulus N and for a large number of different encryption exponents
e. Explain how Eve can use her magic box to try to factor N.

**(b)** Let N = 38749709. Eve’s magic box tells her that the encryption exponent
e = 10988423 has decryption exponent d = 16784693 and that the encryption
exponent e = 25910155 has decryption exponent d = 11514115. Use this
information to factor N.

**(c)** Let N = 225022969. Eve’s magic box tells her the following three encryption/
decryption pairs for N:
(70583995, 4911157), (173111957, 7346999), (180311381, 29597249).
Use this information to factor N.

**(d)** Let N = 1291233941. Eve’s magic box tells her the following three encryption/
decryption pairs for N:
(1103927639, 76923209), (1022313977, 106791263), (387632407, 7764043).
Use this information to factor N.

**Solution:**

**(a)**:

Now, Eve can herself create encryption keys $e_1, e_2, e_3, ..$ as many as she wants and with the help of her magic box she will find the corresponding $d_1, d_2, d_3, ..$ values for them. 

Now we know that given a number K, if $ a^K \equiv a \text{ (mod N) } $ we have 

$$ K \equiv 1 \left(\mod \frac{(p-1)(q-1)}{gcd((p-1),(q-1))} \right) $$

thus for all $i > 0$, we have $d_ie_i - 1$ as a multiple of  $\frac{(p-1)(q-1)}{gcd((p-1),(q-1))}$, thus,  

$ T = gcd( d_1e_1 - 1, d_2e_2 - 1, ...) $ will be a small multiple of  $\frac{(p-1)(q-1)}{gcd((p-1),(q-1))}$.


Now since p-1 and q-1 are even, we can "suppose" that $ T = \frac{ (p-1)(q-1)}{2} $ and then compute $ R = N +1 - (p-1)(q-1) = N +1- 2T$. We want R to be p + q and since we already know N = pq, if we have R as p + q, then we can recover p and q from the quadratic equation $x^2-Rx+N$. If this does not work we continue trying $R = N + 1 - 3T$, $N + 1 -4T$ and so on. When our quadratic equation has two roots, than we can factor p and q. 

Now here, instead of trying different coefficients for T, we can save a bit of time  by changing our method of finding the right multiple of T. Now since we want N + 1 -kT to be equal p+q, and since we know that in practice p and q will have more or less the same order of magnitude. We need k such that $ N + 1 - kT \approx 2\sqrt{N} $ which means that we should take: $ k \approx (N + 1 -2\sqrt{N})/T $

**(b)**:

We have $e_1 = 10988423$, $e_2 = 25910155$, $ d_1 = 16784693$, $ d_2 = 11514115$

we have $ T = gcd(10988423.16784693 - 1, 25910155.11514115 - 1) = 19368558 $

now trying $ R = N + 1 - 1T = 19371152$, however since $x^2 - 19381152x + 38749709$ is irreducable, we need to try with different coefficient for T. Trying $ R = N + 1 - 2T = 12594$ we have $ x^2 -12594x + 38749709 = (x - 7247)(x - 5347)$.

Thus, we have
$ N = 38749709 = 7247 . 5347$.

**(c)**:

We have the encryption/decrytion keys: (70583995, 4911157), (173111957, 7346999), (180311381, 29597249).

\begin{align} T & = gcd(4911157.70583995 - 1, 7346999.173111957- 1, 29597249.180311381-1)  \\ &= gcd(346649081132214, 1271853374967042, 5336720840990868) \\ & = 37498566 \end{align}

Now by the fast calculating method for finding the right multiple of T, Eve calculates:

$$ k \approx (\sqrt{225022969} - 1)^2 / 37498566 \approx 6.00004193 $$

hence, we try R = N + 1 - 6T = 31574, which gives us the polynomial $ x^2 -31574x + 225022969 = (x - 20707)(x - 10867) $.

We have $ N = 20707 . 10867 $

**(d)**:

Similar to (c), 

\begin{align} T & = gcd(76923209.1103927639-1, 106791263.1022313977-1, 7764043.387632407-1) \\ &= gcd(84917656495673550, 109174200786382950, 3009594676141500) \\ & = 129112350 \end{align}


Now by the fast calculating method for finding the right multiple of T, Eve calculates:

$$ k \approx (\sqrt{1291233941} - 1)^2 / 129112350 \approx 10,0002987 $$

thus we have $R = N + 1 - 10T = 110442$ which gives us the polynomial $ x^2 -110442x + 1291233941 = (x - 97151)(x - 13291) $

We have $ N = 97151 . 13291 $



# 3.11

**Question:** 

Here is an example of a public key system that was proposed at a cryptography conference. It was designed to be more efficient than RSA. Alice chooses two large primes $p$ and $q$ and she publishes $N = pq$. It is assumed that $N$ is hard to factor. Alice also chooses three random numbers $g$, $r_1$, and $r_2$ modulo $N$ and computes

$$ g_1 \equiv g^{r_1(p−1)} \text{ (mod N) and } g_2 \equiv g^{r_2(q−1)} \text{ (mod N)}$$.

Her public key is the triple $(N, g_1, g_2)$ and her private key is the pair of primes $(p, q)$.

Now Bob wants to send the message m to Alice, where m is a number modulo N. He chooses two random integers $s_1$ and $s_2$ modulo $N$ and computes

$$ c_1 \equiv mg_1^{s_1} \text{ (mod N) and } c_2 \equiv mg_2^{s_2} \text{ (mod N)}.$$

Bob sends the ciphertext $(c1, c2)$ to Alice.
Decryption is extremely fast and easy. Alice uses the Chinese remainder theorem to solve the pair of congruences

$$ x \equiv c_1 \text{ (mod p) and } x \equiv c_2 \text{ (mod q)} $$

**(a)** Prove that Alice’s solution x is equal to Bob’s plaintext m.

**(b)** Explain why this cryptosystem is not secure.

**Solution:**

**(a)**

Now by Fermat's Little Theorem:

$$ c_1 \equiv mg_1^{s_1} = mg^{s_1r_1(p-1)} \equiv m \mod p \text{ (1)}$$

and 

$$ c_2 \equiv mg_2^{s_2} = mg^{s_2r_2(q-1)} \equiv m \mod q \text{ (2)}$$ 

Thus, combining our solution to (1) and (2), we obtain $ x \equiv m \mod pq $

**(b)**

Although we have $ g_1 \equiv 1 \mod{p} $ we can still (and probably will) get $ g_1 \not\equiv 1 \mod{q}$. From this, we obtain $gcd(g_1 - 1, N) = p$ and using this, Eve can recover p. 

And in case $ g_1 \equiv 1 \mod q$, then our crypted message $ c_1 \equiv m \mod N$, so our message is directly seeable.


# 3.15 (f)

**Question:** Use the Miller–Rabin test to provide a Miller–Rabin witness for the compositeness of n, or conclude that n is probably prime by providing 10 numbers that are not Miller–Rabin witnesses for n.

$$ n = 118901527 $$

**Solution:**

For this question, we are going to write and use an algorithm that tests potential Miller-Rabin witnesses:


In [8]:
# integer n to bested, integer a as potential witness
def MillerRabinTest(n, a):
    # 1
    if (n % 2 == 0 or 1 < gcd(a, n) < n):
        return True # composite
    # 2 write n-1 = 2^{k}q with q odd
    q = n - 1
    k = 0
    while(q % 2 == 0):
        q = q / 2
        k = k + 1
    
    # 3 . Set a = a^{q} mod n
    a = Mod(a, n)^q
    
    # 4. If a is 1 mod n then test fails
    if(Mod(a, n) == 1):
        return False
    
    # 5. 
    for i in range(k):
        # 6.
        if( Mod(a, n) == Mod(-1, n) ):
            return False
        # 7. Set a = a^2 mod n
        a = Mod(a, n)^2
    
    # 9
    return True # composite

So we need to test $n$ with numbers $a$ that are not divisible by $n$. Lets test it with 10 random integers between 1 and n so n cannot divide them.




In [11]:
n = 118901527
mrandom_integers = [choice(range(1, 100)) for i in range(10)]
print("Our randomly choosen integers:", mrandom_integers)

foundWitness = False
for a in mrandom_integers:
    if( MillerRabinTest(n, a)):
        foundWitness = True
        print(a, " is a Miller-Rabin witness for the number ", n)
if( not foundWitness):
    print("None of the randomly choosen 10 numbers is a Miller-Rabin witness for", n)

Our randomly choosen integers: [54, 87, 56, 71, 53, 77, 31, 32, 31, 77]
None of the randomly choosen 10 numbers is a Miller-Rabin witness for 118901527


Thus we can conclude that 118901527 is (probably) a prime number.

# 3.16

**Question:**

Looking back at Exercise 3.10, let’s suppose that for a given N, the magic box can produce only one decryption exponent. Equivalently, suppose that an RSA key pair has been compromised and that the private decryption exponent corresponding to the public encryption exponent has been discovered. Show how the basic idea in the Miller–Rabin primality test can be applied to use this information to factor N.

**Solution:**



# 3.19

**Question:**

We noted in Sect. 3.4 that it really makes no sense to say that the number n has probability 1/ ln(n) of being prime. Any particular number that you choose either will be prime or will not be prime; there are no numbers that are 35% prime and 65% composite! In this exercise you will prove a result that gives a more sensible meaning to the statement that a number has a certain probability of being prime. You may use the prime number theorem (Theorem 3.21) for this problem.

**(a)**

Fix a (large) number N and suppose that Bob chooses a random number n in the interval $\frac{1}{2}N \leq n \leq \frac{3}{2}N$. If he repeats this process many times, prove that
approximately $ 1/ ln(N) $ of his numbers will be prime. More precisely, define:

\begin{align*} P(N) &= \frac{\text{ number of primes between } \frac{1}{2}N \text{ and } \frac{3}{2}N}{\text{ number of integers between } \frac{1}{2}N \text{ and } \frac{3}{2}N} \\ & = \lbrack\text{ Probability that an integer n in the interval } \frac{1}{2}N \leq n \leq \frac{3}{2}N \text{ is a prime number}\rbrack \end{align*}

and prove that:

$$ lim_{N \to ∞} \frac{ P(N)}{ 1/ ln(N)} = 1$$.


**(b)**

More generally, fix two numbers $c_1$ and $c_2$ satisfying $c_1 < c_2$. 

Bob chooses random numbers n in the interval $c_1N \leq n \leq c_2N$. Keeping $c_1$ and $c_2$ fixed, let

$$ P(c_1, c_2, N) = \begin{bmatrix}
   \text{ Probability that an integer n in the interval } \\
   c_1N \leq n \leq c_2N \text{ is a prime number}
\end{bmatrix} $$

In the following formula, fill in the box with a simple function of N so that the statement is true:

$$ \lim_{N \to \infty} \frac{ P(c_1, c_2, N) }{ \boxed{ \text{ } }} = 1 $$

**Solution:**

**(a)**:

If we choose $c_1 = 1/2$ and $c_2 = 3/2$ in part (b), then we have our desired limit as 1. 

**(b)**:

Let $ \pi(X) = (\text{# of primes p satisfying } 2 \leq p \leq X )$

Then we have $ P(c_1, c_2, N) = \frac{\pi(c_2N) - \pi(c_1N)}{N} $

# 3.20



# 3.22(c)

**Question:** Use Pollard’s p − 1 method to factor $ n = 48356747 $

**Solution:** 

By using the pollardsFactorization method that is described belove, we found our two factors as 6917 and 6991.


In [12]:
# input: Integer N to be factored
def pollardsFactorization(N):
    # 1. Set a = 2
    a = 2
    # 2. Loop j = 2,3,4.. up to a specified bound
    j = 2
    while(True):
        # 3. set a = a^j mod N
        a = a^j % N
        # 4. compute gcd(a-1, N)
        d = gcd(a - 1, N)
        # if 1 < d < N, then success
        if(1 < d < N):
            return d
        
        j = j + 1
        
print("Our first factor is:  ", pollardsFactorization(48356747)) # our first factor
print("Our second factor is: ", 48356747/6917 ) # our second factor

Our first factor is:   6917
Our second factor is:  6991


# 3.26

**Question:**
For each part, use the data provided to find values of a and b satisfying $ a^2 \equiv b^2 \text{ (mod N)}$, and then compute $gcd(N, a − b)$ in order to find a nontrivial factor of N

(a) N = 61063

1882^2 ≡ 270 (mod 61063) and 270 = 2 · 3^3 · 5

1898^2 ≡ 60750 (mod 61063) and 60750 = 2 · 3^5 · 5^3

**Solution:**

